# RepeatProfileHMM

A profile-HMM package designed for repeat identification and classification.

# Usage
## `construct`
A program to construct profile-HMM based on the input multiple sequence
alignment result.

### Input
The input must be in [ClustalW format][1]. Both `clustalw2` and `muscle`
can generate this format of multiple alignment.

### Parameters
1. `-u`: turn on heuristic mode
By deault, the MAP is used to determine which columns in the alignment should
be chosen as homologous position, i.e. the model length. This is basically
a dynamic programming algorithm and runs much slower than the other mode,
heuristic mode. However, it generates better model, especially in the case
that most sequences in the alignment show 5' truncation, which is common
in retrotransposons.

2. `-l`: set lambda value
This is the penalizing factor in MAP mode. If `-u` is turned on, it is ignored.
The prbabilistic meaning of this parameter, is the prior probability, in
logarithm, that all columns of the alignment should be included in the model.
Therefore, in most cases, this parameter should be set to a non-positive value.
The bigger this value is, the longer the constructed model will be. 

### Output
The output is a text file which includes: model length, transition probabilities
and emission probabilities. All probabilities are in log scale.

## `scan`
A program to scan the genome using the model file generated by `construct`.

### Input
See `construct`.

### Parameters
1. `-c`: reference genome
Specify the path to a directory of FASTA files of the reference genome.

### Output
A BED file including identified repeats, and the corresponding z-score.

[1]:http://web.mit.edu/meme_v4.9.0/doc/clustalw-format.html
