Wed Feb 22 11:20:30 PST 2017
* `chrF.fa`: a fake chromosome generated by human chrM.
Then 500 repeats were inserted in the genome with lower case
sequences.

* `true_locations.bed`: the insertion sites of simulated copies. All copies
were inserted to the positive strand.

* `repeat_sequences.fa`: the sequences of all inserted repeats, simulated from
one consensus sequence by adding truncation, indel and substitution.

* `repeat_sequences.aln`: the multiple sequence alignment of all copies.

* `repeat_sequences.params`: the model built from the alignment using

    `../bin/construct -o repeat_sequences.params repeat_sequences.aln`

Turn on the -d flag for construct to see which columns are chosen.

* `repeat_sequences.hits`: identification results using

    `../bin/scan -c chrF.fa -o repeat_sequences.hits repeat_sequences.params`

Each region is found by posterior decoding and a posterior z-score
is calculated. Regions with z-score greater than 3 are labeled by "COPY".
