1. `chrF.fa``: a fake chromosome generated by human chrM, presumably no free-of-
repeats. Then 500 repeats were inserted in the genome with lower case
sequences.

2. `true_locations.bed`: the insertion sites of simulated copies. All copies
were inserted to the positive strand.

3. `repeat_sequences.fa`: the sequences of all inserted repeats, simulated from
one consensus sequence by adding truncation, indel and substitution.

4. `repeat_sequences.aln`: the multiple sequence alignment of all copies.

5. `repeat_sequences.params`: the model build from the alignment using

    construct -o repeat_sequences.params repeat_sequences.aln

You can turn on the -d flag for construct to see which columns are
chosen.

6. `repeat_sequences.hits`: identification results using

    scan -c chrF.fa -o repeat_sequences.hits repeat_sequences.params

Each region is found by posterior decoding and a posterior z-score
is given. Regions with z-score greater than 3 is named as a "COPY".
This, however, seems not restrictive enough. You will see many false
positives.
